### PR TITLE
2018-12-28 Understanding Amazon Aurora Endpoints

### PR DESCRIPTION
**Pull request by ahornbeck**
In a comment on the pull request, include the following information and check off the list.

 - [X] Mini bio
Ash Hornbeck is a Sr. Product Architect on the Rackspace Managed Public Clouds Product Architecture Team.  With over 20 years in IT, an MBA, and a BS in Computer Engineering, Ash’s career has included roles ranging from developer to entrepreneur and everything in between.  Over the past several years, he has been focused on helping large enterprises make their journey to the cloud successful.

 - [X] Desired date of publishing
2018-12-28

 - [X] Social media contact info
 https://www.linkedin.com/in/hornbeck/

 - [X] Redaction: (yes|no)
Yes

 - [X] Comments: (yes|no)
Yes

@catlook Please review.

### CODE REVIEW

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/build-63fe6f129b/)

File: _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+First, let's touch on the different endpoints available with Amazon Aurora.
+
+* **Cluster Endpoints** – Cluster Endpoints connect your application to the current primary DB instance for that DB cluster.  Your application can both read and write to this instance.
+
+* **Reader Endpoints** – Reader Endpoints load-balance connections across the pool of available read replicas. Offload read queries here, reducing load on your primary DB instance.
+
+* **Instance Endpoints** – Instance Endpoints connect to a specific instance in the cluster.  Clients can have fine-grained control over query allocation, rather than having Amazon Aurora handle connection distribution.
+
+I know what you're thinking... I'll just connect to the Cluster Endpoint for writes, the Reader Endpoints for all reads, and why would I ever connect to a specific instance – that bypasses the built-in fault tolerance and is asking for trouble, right?  As eluded to earlier, your application and its interaction with Aurora create a complex (or at least a non-trivial) system.  With complex systems, one-size fits all is not a rule to rely on... unless you enjoy calls in the middle of the night.
```
[the remark this answers is not in the document]
Comment (ahornbeck) [reply _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md]: Thanks for the catch... Reader should be singular.  I'll fix it.

File: _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+Let's go through some scenarios and look at when and where to use the various endpoints.
+
+## Immediate Consistency
+
+Some applications expect the data to be immediately consistent.  When these applications write data, they immediately read the data in strict compliance with any of the many design patterns that say, "rely on the model, not your local data".  Amazon Aurora is ACID-compliant; a read from the Cluster Endpoint immediately after a successful write commit will retrieve the expected data (assuming another transaction did not modify the data before the subsequent read).
+
+Where the trouble happens is when writes are sent to the Cluster Endpoint and reads are made to the Reader Endpoints.  This is due to the latency between writing data and when it becomes visible to the readers.  While the replication latency is under 100 milliseconds, it isn't instantaneous, leading to a race condition.  If you have a scenario where you need to immediately read data after writing, use the Cluster Endpoint for both reads and writes.  It should be noted, if additional performance is required, you must increase the size of your primary instance.
```
[the remark this answers is not in the document]
Comment (ahornbeck) [reply _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md]: 👍

File: _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+authorAvatar: https://www.gravatar.com/avatar/f96915d3f29e5623f9a9ff8c7cb3148f
+bio: "Ash Hornbeck is a Sr. Product Architect on the Rackspace Managed Public Clouds Product Architecture Team."
+categories:
+    - AWS
+    - Database
+metaTitle: "Understanding Amazon Aurora Endpoints"
+metaDescription: "A look at the various Aurora Endpoint types and when to use them."
+ogTitle: "Understanding Amazon Aurora Endpoints"
+ogDescription: "A look at the various Aurora Endpoint types and when to use them."
+---
```
[the remark this answers is not in the document]
Comment (ahornbeck) [reply _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md]: Done.

File: _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+## Offloading Reads
+
+If immediate reads after writes are not ok, what's the point of having a separate Reader Endpoint?  There are many use-cases where slightly inconsistent data is not an issue.  For  example: daily reporting.  When running batch jobs to generate reports from yesterday's data, a 100-millisecond replication lag is inconsequential.  It's easy to imagine other scenarios, such as product descriptions on an ecommerce site... you are more likely to have stale data in the user's browser than be impacted by the replication lag.
+
+Generally, read-heavy workloads that do not rely on immediate consistency should consider using the Reader Endpoint.
+
+## Non-uniform Workloads
+
+There are situations where it makes sense to use Instance Endpoints.  Let's say you have an application that has a number of stateless microservices behind an Elastic Load Balancer.  In this situation, it is reasonable to assume that queries will be evenly distributed across the clients, and therefore, the Read Replicas – when using the Reader Endpoint.
```
[the remark this answers is not in the document]
Comment (ahornbeck) [reply _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md]: thank you.

File: _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+But what about the situation where the workload of each client is not uniformly distributed?  If a reporting service is added to the mix, one Read Replica will have a disproportionally high load as compared to the others.  If the before-mentioned microservices also use this Read Replica, there is a high probability of poor query performance when querying the instance used by the reporting service.  One approach to avoid this is to manage the distribution of connections on the client side, rather than letting Amazon Aurora do this for you.  Luckily, this is easily done by using Instance Endpoints instead of the Reader Endpoint.  With the MariaDB Connector/J for Aurora for MySQL or Fast Failover with Amazon Aurora PostgreSQL, the driver can be made aware of the individual instances you want to use as read replicas, allowing the driver to directly manage how queries are distributed to individual instances.
```
[the remark this answers is not in the document]
Comment (ahornbeck) [reply _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md]: going with all capitalized.

File: _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+## Manage DNS Caching
+
+In addition to understanding the nature of your workloads, it is important to understand how connections are allocated to provide Aurora's High Availability and Reader Endpoint load balancing.
+
+Management of automatic Cluster Endpoint failover and Reader Endpoint load balancing are handled through DNS (Amazon Route 53), not at the IP, TCP, or database client protocol layers. Handling distribution of connections through DNS means your application must make a DNS query for each new connection request.  Applications that use DNS caching should adjust the cache timeout to match the DNS record TTL for Amazon Aurora.  Caching DNS responses longer than the Aurora DNS record TTL will result in a couple of error conditions.  If there is a High-Availability (HA) failover event, that is, a Read Replica is promoted to Primary, a cached DNS response will mean your application will attempt to reconnect to the old, failed instance.  In the case of the Reader Endpoint, caching the DNS response will result in multiple connections going to a specific Read Replica instead of being distributed across the available Read Replicas.
```
[the remark this answers is not in the document]
Comment (ahornbeck) [reply _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md]: thank you.

File: _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+## Wrapping Up
+
+As we can see, there is no one-size-fits-all solution when it comes to a non-trivial production workload.  If your application needs immediate consistency, then make sure both reads and writes are sent to the Cluster Endpoint.  When read queries can handle a little replication lag, offload read queries to the Reader Endpoints.  If you can use Read Replicas but your queries are not uniformly distributed across all clients, then use the Instance Endpoints and manage the distribution of connections on the client side.  Just don't forget, if your application uses DNS caching with cache timeouts longer than the DNS record TTL, endpoints will appear to not behave as expected during Cluster Endpoint HA failovers or when using the Reader Endpoint.
+
+Understanding the nature of your application, how the Aurora Endpoints behave, and correctly applying that knowledge will lead to a more robust application environment, a very important thing to you and your customers.  No one wants an outage call in the middle of the night.
```
[the remark this answers is not in the document]
Comment (ahornbeck) [reply _posts/2018-12-28-Understanding-Amazon-Aurora-Endpoints.md]: thank you.

**Comment by ahornbeck** [comment]
@catlook - Thanks for the fixed.  They have been applied.

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-a07ed6c63d/).

**Pull request merged**